### PR TITLE
Add stream preview element to generic camera

### PR DESCRIPTION
**Pull request by davet2001**
## Proposed change

As part of a [PR to enable a more user-friendly experience when configuring video stream sources](https://github.com/home-assistant/core/pull/122563), a preview element is added for previewing the still image and stream from a generic camera during configuration.

This is done by building on the `FlowPreviewGeneric` element, which had to be slightly modified to allow it to be subclassed.
Also included is an svg spinner displayed by the video element while the stream is loading, from https://github.com/n3r4zzurr0/svg-spinners

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion: https://github.com/home-assistant/core/pull/122563
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling for playlist URLs in the media player, improving robustness.
	- Added support for a new preview type, "generic_camera," enhancing the preview system.
	- Introduced a new component for rendering camera feeds, allowing display of still images and live streams.

- **Improvements**
	- Increased modularity and extensibility of the preview components, facilitating better integration and functionality.

### CODE REVIEW

**Comment by davet2001** [comment]
@balloob is this in line with what you meant in your [comment on #21246](https://github.com/home-assistant/frontend/pull/21246#pullrequestreview-2161551260)?

File: src/dialogs/config-flow/previews/flow-preview-generic_camera.ts
Comment (bramkragten) [inline src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:27]:
```suggestion
```

Comment (bramkragten) [reply src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:27]: yes, the default is false, but also because it is a boolean, you can set it to true by just adding the attribute like you did for `playsinline` for example.

Comment (davet2001) [reply src/dialogs/config-flow/previews/flow-preview-generic_camera.ts]: Thanks!

**Comment by davet2001** [comment]
@bramkragten 
Thanks for the review, it is working ok but I get this client side console error whenever I tick/untick the box to accept the preview:
```
Calling service system_log write 
Object { logger: "frontend.js_dev.modern.202407190", message: 'Unhandled promise rejection from Firefox 128.0 on Ubuntu\n{\n  "code": "not_found",\n  "message": "Subscription not found."\n}', level: "debug" }
 undefined [connection-mixin.ts:94](webpack://home-assistant-frontend/src/state/connection-mixin.ts?5cdf)
Uncaught (in promise) 
Object { code: "not_found", message: "Subscription not found." }
[connection.js:313](webpack://home-assistant-frontend/node_modules/home-assistant-js-websocket/dist/connection.js?f1aa)
    unsubscribe connection.js:313
    AsyncFunctionThrow self-hosted:808
    (Async: async)
    subscribeMessage connection.js:327
    _subscribePreview flow-preview-generic.ts:74
    InterpretGeneratorResume self-hosted:1417
    AsyncFunctionNext self-hosted:804
    (Async: async)
```
Any ideas on how to deal with this?

**Comment by davet2001** [comment]
I have tried rebasing this but it is no longer working correctly. Will take a little longer to figure out. Setting to draft for now.

**Comment by davet2001** [comment]
Now working again, after url was re-added to `<ha-hls-player>`. Ready for review.

**Comment by davet2001** [comment]
I am not 100% sure that this PR should have the 'wait for backend' tag.  The frontend changes can be made safely without the backend, as they just create lit elements that are not used until the backend is changed.

It is more the backend changes that can only be made once the frontend is changed.

**Comment by bramkragten** [comment]
> I am not 100% sure that this PR should have the 'wait for backend' tag. The frontend changes can be made safely without the backend, as they just create lit elements that are not used until the backend is changed.
> 
> It is more the backend changes that can only be made once the frontend is changed.

We don't merge code until both frontend and backend are approved to avoid the backend changing after the frontend is merged or the other way around. We also dont want unused code in the project.

**Comment by davet2001** [comment]
Ok, understood. Both are now ready from a test coverage point of view.

**Comment by allenporter** [comment]
The core pr is approved, removing waiting for backend label.

File: src/dialogs/config-flow/previews/flow-preview-generic_camera.ts
Comment (bramkragten) [inline src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:29]: Do we need to add a spinner as the poster?

I'd rather show a normal loader spinner we want to do this and remove it on the `load` event of `ha-hls-player`

File: src/dialogs/config-flow/previews/flow-preview-generic_camera.ts
Comment (bramkragten) [inline src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:9]: I think this should be:
```suggestion
    if (!this._error) {
```
?

Comment (davet2001) [inline src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:9]: The line is acting as a guard for 
`this._preview.xxxx` below it.
How about:
```suggestion
    if (this._error || !this._preview) {
```

Comment (bramkragten) [reply src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:9]: That will show an error while loading, not what we want.
You can add this if you just want to guard:

```ts
if (!this._preview) {
	return nothing;
}
```

Or return a spinner or something

Comment (davet2001) [reply src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:10]: Agreed. Done 447519f

Comment (davet2001) [reply src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:10]: Ok. done 447519f

File: src/dialogs/config-flow/previews/flow-preview-generic_camera.ts
Comment (davet2001) [reply src/dialogs/config-flow/previews/flow-preview-generic_camera.ts]: Ok. Done 447519f.

**Comment by davet2001** [comment]
@silamon Maybe I don't know the process fully, but the backend PR is not merged yet.

File: src/dialogs/config-flow/previews/flow-preview-generic_camera.ts
Comment (bramkragten) [inline src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:48]:
```suggestion
  private _videoLoaded() {
      this.shadowRoot!.getElementById("hls-load-spinner")?.remove();
  }
}
```

Comment (bramkragten) [reply src/dialogs/config-flow/previews/flow-preview-generic_camera.ts:48]: Do we need to reshow the spinner every time the `preview` changes?

**Review by bramkragten — DISMISSED** [review]

**Review by silamon — APPROVED** [review]

**Pull request merged**
